### PR TITLE
fix(core): ignore CNAME entries in node:dns resolved-address validation

### PR DESCRIPTION
**Pull request by LJAYi**
## Summary

- fix `node:dns` resolved-address validation on Cloudflare Workers by ignoring the CNAME entries workerd's DoH-based `lookup()` returns
- the resolved-address preflight stays **active** on Workers — it is corrected, not disabled
- fail closed when filtering leaves no address to validate
- Node/Docker deployments keep the full DNS fail-closed guard unchanged
- add regression tests for the CNAME filtering and the fail-closed path

## Problem

After upgrading a Cloudflare Workers deployment to v1.2.0, saving credentials for multiple unrelated public providers fails before the request reaches the provider:

```text
request URL must not resolve to private or reserved IP addresses
```

Observed examples:

- OOMOL Fusion API credential validation at `https://fusion-api.oomol.com/openapi/qwen-image?hideTaskStateAPI=true`
- Tailscale OAuth client validation at `https://api.tailscale.com/api/v2/oauth/token`

Both names resolve publicly, yet the guard rejected them. Hosts with plain A records (e.g. `api.github.com`) resolved cleanly, which is why only some providers broke.

## Root cause

PR #97 (closing #94) made `providerFetch` dynamically import `node:dns/promises` and validate every resolved address, expecting that import to be unavailable on runtimes such as Cloudflare Workers.

The Worker configuration enables `nodejs_compat`, so `node:dns` **is** importable there (both `dns` and `dns/promises` are in `@cloudflare/unenv-preset`'s `defaultNativeModules`). workerd's `lookup()` resolves over DoH and returns the host's genuine public addresses — but it maps *every* DoH answer record into an entry without filtering by record type. A CNAME answer therefore arrives as `{ address: "controlplane.tailscale.com.", family: 4 }` alongside the real A/AAAA records.

`isBlockedIpAddress` treats unparseable input as blocked, so a single CNAME entry made the guard reject every CNAME'd host — `api.tailscale.com`, `fusion-api.oomol.com`, `graph.microsoft.com`, `api.dropboxapi.com` — before any provider request could run.

## Fix

Filter non-IP entries in the default lookup adapter (`resolveDefaultLookup` in `src/core/guarded-fetch.ts`), keeping only real A/AAAA records via the new `isIpAddress` helper (`src/core/request.ts`):

```ts
const addresses = results.filter((entry) => isIpAddress(entry.address));
if (addresses.length === 0) {
  throw new Error(`${hostname} did not resolve to any IP address`);
}
```

The real addresses are always present alongside the CNAME entries, so this clears the false positive without giving up the DNS name → private-IP defense added in #97 (which AGENTS.md requires for credential-derived egress hosts).

It fails closed when filtering leaves nothing: returning an empty list would let the check pass vacuously, so a resolver coaxed into answering with only CNAMEs could otherwise skip validation while the transport resolves the name to a blocked address on its own. Throwing is reported by `assertResolvedAddressesAllowed` as "could not be resolved for validation", matching a lookup failure.

The change lives in the shared default adapter, so it applies to every deployment target — the resolved-address preflight keeps working on Cloudflare Workers rather than being turned off there.

## Security boundary

- resolved-address validation remains **enabled** on all targets, Workers included
- literal URL SSRF validation (loopback, private, link-local, metadata, reserved, multicast, IPv6) unchanged
- redirect `Location` validation and credential-header stripping on cross-origin redirects unchanged
- Node/Docker deployments retain the full DNS-rebinding defense added in #97
- this change does **not** enable `OOMOL_CONNECT_ALLOW_PRIVATE_NETWORK`

## Verification

Verified on workerd (`nodejs_compat`, compatibility date 2026-07-02):

```text
host                    main      this
api.tailscale.com       BLOCKED   ALLOWED
fusion-api.oomol.com    BLOCKED   ALLOWED
api.github.com          ALLOWED   ALLOWED
localtest.me (->127.1)  BLOCKED   BLOCKED
```

Entries kept after filtering:

```text
host                   entries  IPs kept  non-IP
api.tailscale.com      34       32        2
fusion-api.oomol.com    4        2        2
graph.microsoft.com    16       12        4
api.dropboxapi.com      6        2        4
api.github.com          1        1        0
(nonexistent)          throws ENOTFOUND
```

- `vitest run`
- `oxlint .`
- `oxfmt --check .`
- `node scripts/typecheck.ts src scripts-all examples`

---

> [!NOTE]
> **Corrected after merge (see #145).** The original title and body described a Workers-only "disable the preflight" approach — `setDefaultGuardedFetchDnsLookup(null)` in the Cloudflare entry point, with a test injecting `198.18.0.1`. That approach was replaced during review by the CNAME-filtering fix that actually merged in `7ab3d3e` (`src/core/guarded-fetch.ts` + `src/core/request.ts`); `src/server/cloudflare.ts` was never touched. The squash-merge preserved the outdated text as the commit message, so this description has been rewritten to match the code.

### CODE REVIEW

**Comment by BlackHole1** [comment]
Thanks for chasing this down — the localization was right (Workers + the `node:dns` preflight, on the shared `providerFetch` path), and the Tailscale `skipDnsValidation` observation is a good catch. I pushed a commit to this branch that keeps the diagnosis but changes the remedy, because measuring workerd directly turned up a different root cause.

## workerd's DNS is not returning proxy addresses

I ran this repo's own pinned workerd (`wrangler dev`, `nodejs_compat`, compat date 2026-07-02) and called `lookup(host, { all: true })`:

```
api.tailscale.com    → [{address:"controlplane.tailscale.com.",family:4}, {address:"192.200.0.108",family:4}, ...,
                        {address:"controlplane.tailscale.com.",family:6}, {address:"2606:b740:49::115",family:6}, ...]
fusion-api.oomol.com → [{address:"alb-2ez87aepnoql6znr3g.ap-southeast-1.alb.aliyuncsslbintl.com.",family:4},
                        {address:"8.219.95.213",family:4}, {address:"8.219.110.135",family:4}, ...]
example.com          → [{address:"104.20.23.154",family:4}, ...]   // all real IPs, no CNAME
```

Those are the **correct public addresses** — the same `8.219.x.x` / `192.200.0.x` / `2606:b740:49::` your description cites from public resolvers. workerd resolves over DoH and gets it right (it even bypasses a local fake-IP resolver).

The actual defect is visible in workerd's own implementation:

```js
sendDnsRequest(hostname, 'A').catch(() => ({ Answer: [] })),
const ipv4Addresses = ipv4Response.Answer?.map((answer) => ({ address: answer.data, family: 4 }))
```

No `answer.type` filter — so **CNAME (type 5) answers get mapped as addresses**, putting a hostname string in `address`. `isBlockedIpAddress` treats unparseable input as blocked (deliberate fail-closed, `request.ts`), so a single CNAME entry throws exactly the error you reported.

**The failure predicate is "the host has a CNAME"**, which is why only some providers broke:

| broken (CNAME) | fine (plain A) |
|---|---|
| `api.tailscale.com`, `fusion-api.oomol.com`, `graph.microsoft.com`, `api.dropboxapi.com` | github, notion, openai, slack, box, hubapi, airtable |

Microsoft Graph and Dropbox were hit too, beyond the two in the description.

## Why the remedy changed

Since the addresses are trustworthy and only the interleaved CNAME entries aren't, disabling the layer gives up a control that works. The real A/AAAA records are always present alongside the CNAME entries, so filtering non-IP entries fixes the false positive and keeps the DNS name→private-IP defense from #97 — which `AGENTS.md` specifically requires for credential-derived egress hosts:

> NEVER add it when the host comes from credential/user input ... there the DNS check is the SSRF defense, not redundant overhead.

Three-way comparison, all measured on real workerd (stub transport, so the only variable is the guard):

| host | main today | this PR (disable) | pushed commit |
|---|---|---|---|
| `api.tailscale.com` | BLOCKED ← the bug | ALLOWED | ALLOWED |
| `fusion-api.oomol.com` | BLOCKED ← the bug | ALLOWED | ALLOWED |
| `api.github.com` | ALLOWED | ALLOWED | ALLOWED |
| `localtest.me` → `127.0.0.1` | BLOCKED | **ALLOWED** | BLOCKED |

That last row is the reason. `localtest.me` is an ordinary public hostname that resolves to loopback, so `assertPublicHttpUrl` passes it and only the DNS check stops it.

On the "native Workers networking as the final boundary" point: under `wrangler dev` I fetched a loopback HTTP server from inside workerd and got **200 with the body** for both `http://127.0.0.1:PORT/` and `http://localtest.me:PORT/`. So the runtime doesn't provide that backstop on the local dev path (`npm run dev:cloudflare`). Production Cloudflare egress may well block RFC1918, but I couldn't verify it, and it seemed better not to rest the guard on an unverified platform behavior.

## What's in the commit

- `guarded-fetch.ts`: filter non-IP entries in the default `node:dns` adapter; corrected the docblock that claimed Workers lacks `node:dns` (that assumption is what let this reach production)
- `request.ts`: exported `isIpAddress`
- tests: the fixture is workerd's **real** output including the CNAME entries, so it reproduces the actual bug — reverting the filter makes it fail with `request URL must not resolve to private or reserved IP addresses`. Plus a case proving a CNAME'd host resolving to `169.254.169.254` is still blocked.

I dropped `cloudflare-dns.test.ts` and the `cloudflare.ts` change since the entry point no longer needs to opt out. `vitest run` 522 passed, `oxlint`, `oxfmt --check`, and typecheck across all four tsconfigs are clean.

Worth reporting `lookup()`'s missing `answer.type` filter upstream to workerd — it'll bite anyone validating resolved addresses on Workers.

File: src/core/guarded-fetch.ts [hunk 1]
```diff
@@ -295,8 +296,17 @@ async function resolveDefaultLookup(): Promise<GuardedFetchDnsLookup | null | un
...
+        const results = await lookup(hostname, { all: true });
+        // Keep only real addresses. workerd's node:dns resolves over DoH and maps
+        // every answer record into an entry without filtering by record type, so a
+        // CNAME answer arrives as { address: "target.example.com.", family: 4 }.
+        // isBlockedIpAddress treats unparseable input as blocked, which would
+        // reject every CNAME'd host (api.tailscale.com, graph.microsoft.com, ...).
+        // The real A/AAAA records are present alongside, so dropping non-addresses
+        // keeps the resolved-address check intact rather than disabling it.
+        return results.filter((entry) => isIpAddress(entry.address));
+      },
```
[the remark this answers is not in the document]
Comment (BlackHole1) [reply src/core/guarded-fetch.ts:317]: Good catch — fixed in 5b032eb. The concern is real and I'd talked myself out of it: I reasoned that a dangling CNAME wouldn't resolve for the transport either, but that only holds if both resolvers see the same answer. Since the attacker controls their own zone, "DoH sees CNAME-only, Cloudflare's fetch resolver sees an A record" is constructible, which is exactly the forced-failure / split-resolver bypass the lookup-rejection path already fails closed against. Passing an empty list to the validation loop made the check pass vacuously.

Took the suggestion with two adjustments:

- The message is `${hostname} did not resolve to any IP address`, since `assertResolvedAddressesAllowed` catches anything the lookup throws and re-reports it as `"request URL could not be resolved for validation"` via `createError` — so it lands as a `ProviderRequestError(502)` and matches a lookup failure, rather than surfacing a raw `Error`.
- The comment records *why* it fails closed rather than what the line does.

Before taking it I checked the obvious risk — that the guard reintroduces the false positive this PR fixes. Measured on real workerd (`nodejs_compat`, compat date 2026-07-02):

| host | entries | IPs kept | non-IP |
|---|---|---|---|
| `api.tailscale.com` | 34 | 32 | 2 |
| `fusion-api.oomol.com` | 4 | 2 | 2 |
| `graph.microsoft.com` | 16 | 12 | 4 |
| `api.dropboxapi.com` | 6 | 2 | 4 |
| `api.github.com` | 1 | 1 | 0 |
| nonexistent host | — | — | throws `ENOTFOUND` |

Every affected host keeps real addresses after filtering, so `length === 0` is unreachable for them. NXDOMAIN throws before the filter, so the existing catch already handled it — the only way to hit the empty case is the CNAME-only answer you described.

Also took the nitpick: added a fail-closed test covering both a CNAME-only answer and an empty result, asserting the request is rejected and the transport is never invoked. Mutation-checked it — removing the guard makes it fail.

`vitest run` 523 passed, `oxlint`, `oxfmt --check`, and typecheck across all four tsconfigs clean.

**Comment by LJAYi** [comment]
@BlackHole1 Thank you for the excellent follow-up on this fix. The direct workerd measurements, identification of the interleaved CNAME records, and the fail-closed handling for empty filtered results made the final solution both more precise and more secure than the original workaround. I really appreciate the careful investigation and the detailed explanation.

Would it be possible to publish a patch release containing this merged fix? The current latest release is still v1.2.0, so Cloudflare Workers deployments tracking stable releases continue to fail when connecting CNAME-backed providers such as Tailscale and Fusion API. Our production deployment intentionally syncs immutable upstream releases rather than upstream main, so a patched release would let it receive the fix cleanly through the normal update path. Thanks again!

**Comment by BlackHole1** [comment]
@LJAYi Done
